autonat/autonat-v2.md

```diff
@@ -0,0 +1,207 @@
+
+`E_DIAL_REFUSED`: the peer didn't attempt a dial because of rate limiting,
+resource limit reached or blacklisting.
+
+`E_TRANSPORT_NOT_SUPPORTED`: the peer didn't have the ability to dial any of the
+requested addresses.
+
+`E_BAD_REQUEST`: the peer didn't attempt a dial because it was unable to decode
+the message.
+
+`E_INTERNAL_ERROR`: error not classified within the above error codes occured on
+peer that prevented it from completing the request.
+
+Implementations MUST discard responses with status codes they do not understand
+
+### Consideration for DDOS Prevention
+
+In order to prevent attacks like the one described in [RFC 3489, Section
+12.1.1](https://www.rfc-editor.org/rfc/rfc3489#section-12.1.1) (see excerpt
+below), implementations MUST NOT dial any multiaddress unless it is based on the
+IP address the requesting node is observed as. This restriction as well implies
+that implementations MUST NOT accept dial requests via relayed connections as
+one can not validate the IP address of the requesting node.
+
+> RFC 3489 12.1.1 Attack I: DDOS Against a Target
+>
+> In this case, the attacker provides a large number of clients with the same
+> faked MAPPED-ADDRESS that points to the intended target. This will trick all
+> the STUN clients into thinking that their addresses are equal to that of the
+> target. The clients then hand out that address in order to receive traffic on
+> it (for example, in SIP or H.323 messages). However, all of that traffic
+> becomes focused at the intended target. The attack can provide substantial
+> amplification, especially when used with clients that are using STUN to enable
+> multimedia applications.
+
+
+## Implementation Suggestions
+
+For any given address, implementations SHOULD do the following
+- periodically recheck reachability status
+- query multiple peers to determine reachability
+
+The suggested heuristic for implementations is to consider an address reachable
+if more than 3 peers report a successful dial and to consider an address
+unreachable if more than 3 peers report unsuccessful dials. 
+
+Implementations are free to use different heuristics than this one
+
+
+## RPC Messages
+
+All RPC messages sent over a stream are prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec]. 
+
+All RPC messages on stream `/libp2p/autonat/2.0.0/dial` are of type
+`DialMessage`. A `DialRequest` message is sent as a `DialMessage` with the `dialRequest`
+field set and the `type` field set to `DIAL_REQUEST`. `DialResponse` is handled
+similarly.
+
+On stream `/libp2p/autonat/2.0.0/attempt`, there is a single message type
+`AttemptMessage`
+
+```proto
+syntax = "proto3";
+
+message DialMessage {
+  enum Type {
+    DIAL_REQUEST  = 0;
+    DIAL_RESPONSE = 1;
+  }
+
+  Type type          = 1;
+  DialRequest dialRequest   = 2;
+  DialResponse dialResponse = 3;
```

Comment:
```suggestion
  optional DialRequest dialRequest   = 2;
  optional DialResponse dialResponse = 3;
```

The proto3 presence rules still confuse me. From what I understand we should be introducing explicit presence here, no?

See also https://github.com/libp2p/specs/issues/465 for details.

Reply:
I think it's not required for message types. Please correct me if I'm wrong. 
 
this shouldn't be a problem because the default value of a message type should be NULL. For ex in go this would translate to a pointer to DialRequest ```*DialRequest``` and pointer to DialResponse whose value will be nil if it is not set. The message also has a Type field which signals which field to check 

```msg.Type == DIAL_REQUEST``` => ```msg.dialRequest``` should not be nil 

I'm fine with changing this if it simplifies implementation in other languages.